book/src/dev/rfcs/0010-v5-transaction.md

```diff
@@ -0,0 +1,262 @@
+
+```rust
+struct Spend<AnchorV: AnchorVariant> {
+    cv: commitment::ValueCommitment,
+    anchor: AnchorV::PerSpend,
+    nullifier: note::Nullifier,
+    rk: redjubjub::VerificationKeyBytes<SpendAuth>,
+    zkproof: Groth16Proof,
+    spend_auth_sig: redjubjub::Signature<SpendAuth>,
+}
+```
+
+### No Changes to Sapling Output
+[no-changes-to-sapling-output]: #no-changes-to-sapling-output
+
+In Zcash the Sapling output representations are the same for V4 and V5 transactions, so no variants are needed. The output code is located at `zebra-chain/src/sapling/output.rs`:
+
+```rust
+struct Output {
+    cv: commitment::ValueCommitment,
+    cm_u: jubjub::Fq,
+    ephemeral_key: keys::EphemeralPublicKey,
+    enc_ciphertext: note::EncryptedNote,
+    out_ciphertext: note::WrappedNoteKey,
+    zkproof: Groth16Proof,
+}
+```
+
+## Orchard Additions
+[orchard-additions]: #orchard-additions
+
+### Adding V5 Transactions
+[adding-v5-transactions]: #adding-v5-transactions
+
+Now lets see how the V5 transaction is specified in the protocol, this is the second table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus) and how are we going to represent it based in the above changes for Sapling fields and the new Orchard fields.
+
+We propose the following representation for transaction V5 in Zebra:
+
+```rust
+enum Transaction::V5 {
+    lock_time: LockTime,
+    expiry_height: block::Height,
+    inputs: Vec<transparent::Input>,
+    outputs: Vec<transparent::Output>,
+    sapling_shielded_data: Option<sapling::ShieldedData<SharedAnchor>>,
+    orchard_shielded_data: Option<orchard::ShieldedData>,
+}
+```
+
+To model the V5 anchor type, `sapling_shielded_data` uses the `SharedAnchor` variant located at `zebra-chain/src/transaction/sapling/shielded_data.rs`.
+
+### Adding Orchard ShieldedData
+[adding-orchard-shieldeddata]: #adding-orchard-shieldeddata
+
+The new V5 structure will create a new `orchard::ShieldedData` type. This new type will be defined in a separated file: `zebra-chain/src/orchard/shielded_data.rs` and it will look as follows:
+
+```rust
+struct orchard::ShieldedData {
+    flags: Flags,
+    value_balance: Amount,
+    anchor: tree::Root,
+    proof: Halo2Proof,
+    /// An authorized action description.
+    ///
+    /// Storing this separately ensures that it is impossible to construct
+    /// an invalid `ShieldedData` with no actions.
+    first: AuthorizedAction,
+    rest: Vec<AuthorizedAction>,
+    binding_sig: redpallas::Signature<redpallas::Binding>,
+}
+```
+
+The fields are ordered based on the **last** data deserialized for each field.
+
+### Adding Orchard AuthorizedAction
+[adding-orchard-authorizedaction]: #adding-orchard-authorizedaction
+
+In `V5` transactions, there is one `SpendAuth` signature for every `Action`. To ensure that this structural rule is followed, we create an `AuthorizedAction` type:
```

Comment:
This is true if there are spends / if the enableSpends flag is true. But it's not always true. There may be Actions that do not spend, like there are Outputs in Sapling. So we might need `Either<Action, AuthorizedAction>` or something like that in `orchard::ShieldedData`

Reply:
@teor2345

Reply:
Ah, so to be clear, what's the precise requirement?
1. If enableSpends is set, there must be an Authorization
2. If there are any non-dummy spends, there must be an Authorization

Because that affects the data modelling.

Reply:
The spec says that there are always the same number of `Action`s and `SpendAuth`s.
![image](https://user-images.githubusercontent.com/8951843/112069569-cc6e7880-8bb7-11eb-837a-12c4c54ae668.png)

And that dummy notes are allowed:
![image](https://user-images.githubusercontent.com/8951843/112072155-f5ddd300-8bbc-11eb-8dba-879ab81001c5.png)
![image](https://user-images.githubusercontent.com/8951843/112072172-fe360e00-8bbc-11eb-864a-fa8be2308ef0.png)

So any spend or output might be a dummy, but:
* if we can tell without the flags, that seems like a metadata leak security issue
* if we can tell using the flags, that seems like we might want to use an enum instead of bitflags.

Note: GitHub has hidden this conversation in the files view, because of another overlapping change. 😞

Reply:
Here are some possible ideas, but I don't think we should use them without more revisions.

`Disabled` encoded into types:
```rust
enum orchard::TransactionAction {
    SpendAndOutputEnabled {
        action: Action,
        spend_auth_sig: redpallas::Signature<redpallas::SpendAuth>,
    }
    OutputDisabled {
        action: Disabled<Action, Output>,
        spend_auth_sig: redpallas::Signature<redpallas::SpendAuth>,
    }
    SpendDisabled {
        action: Disabled<Action, Spend>,
       spend_auth_sig: Disabled<redpallas::Signature<redpallas::SpendAuth>>,
    }
    OutputAndSpendDisabled {
        action: Disabled<Action, Spend + Output>,
        spend_auth_sig: Disabled<redpallas::Signature<redpallas::SpendAuth>>,
    }
}
```

We could make the types easier to understand if:
* we could ignore dummy data (but we can't, because it's consensus-critical)
* we could split the `Action` into `Spend`, `Output`, and `Common` fields (but I'm not sure if we want to diverge from `zcashd`)

Here's how it looks with the spend and output split:
```rust
enum orchard::TransactionAction {
    SpendAndOutputEnabled {
        cv: commitment::ValueCommitment,
        spend: ActionSpend,
        output: ActionOutput,
        spend_auth_sig: redpallas::Signature<redpallas::SpendAuth>,
    }
    OutputDisabled {
        cv: commitment::ValueCommitment,
        spend: ActionSpend,
        output: Disabled<ActionOutput>,
        spend_auth_sig: redpallas::Signature<redpallas::SpendAuth>,
    }
    SpendDisabled {
        cv: commitment::ValueCommitment,
        spend: Disabled<ActionSpend>,
        output: ActionOutput,
        spend_auth_sig: Disabled<redpallas::Signature<redpallas::SpendAuth>>,
    }
    OutputAndSpendDisabled {
        cv: Disabled<commitment::ValueCommitment>,
        spend: Disabled<ActionSpend>,
        output: Disabled<ActionOutput>,
        spend_auth_sig: Disabled<redpallas::Signature<redpallas::SpendAuth>>,
    }
}
```

Reply:
Dummy notes are not visible at all at the consensus layer; you don't need to consider them in data modeling.

The `enableSpendsOrchard` and `enableOutputsOrchard` flags are supposed to be the minimal possible mechanism to allow forcing the shielded input or shielded output values to zero. They don't remove anything from the data model, and the only interaction they have with consensus outside the circuit, in NU5, is that the `enableSpendsOrchard` flag, if present, MUST be 0 for coinbase transactions. You shouldn't (IMHO) be implementing different code paths for those flags, because that is only likely to lead to bugs.

Reply:
> The `enableSpendsOrchard` and `enableOutputsOrchard` flags are supposed to be the minimal possible mechanism to allow forcing the shielded input or shielded output values to zero in the circuit. ... they don't affect how spend authorization signatures or proofs or Action descriptions are validated; and the only interaction they have with consensus outside the circuit, in NU5, is that the `enableSpendsOrchard` flag, if present, MUST be 0 for coinbase transactions.

From reading the spec, I see that they're just inputs to the proof circuit, which handles all the validation:

<img width="870" alt="Screen Shot 2021-03-24 at 09 13 27" src="https://user-images.githubusercontent.com/8951843/112230801-3a32a700-8c81-11eb-8489-1fb3e78d7433.png">

Based on this information, I don't think we need to make any changes to the current design.